include/functions/math.hpp

```diff
@@ -125,6 +124,60 @@ BigInt sqrt(const BigInt& num) {
     return sqrt_current;
 }
 
+/*
+   is_probable_prime
+   -----------------
+   Returns bool that indicates the primality of a BigInt based 
+   on primality testing using Rabin-Miller testing
+   NOTE: certainty determines how accurate test is and must be greater than 0
+   			
+ */
+
+bool BigInt::is_probable_prime(size_t certainty) const {
+    BigInt num = this->value;
+
+    // Covering corner cases
+    if (num <= 1 or num == 4) return false;
+    if (num <= 3) return true;
+    
+    // Find r such that n = 2^d * r + 1 for some r >= 1
+    BigInt d = num - 1;
+    while (d % 2 == 0) {
+        d /= 2;    
+    }
+    
+    BigInt temp_d = d;
+    BigInt temp2_d = d;
+    BigInt TWO = 2;
+    while (certainty-- > 0) {
+
+        // Corner cases ensure num > 4
+        BigInt a =  TWO + BigInt(rand()) % (num - 4);
+        BigInt res = 1;
+        BigInt x = a % num;
+        while (temp2_d > 0) {
+            // Getting last character from BigInt and checking if even or odd
+            std::string temp_s = temp2_d.to_string();
+            if ((int) temp_s.back() & 1) {
```

Comment:
We could use this version but there will be a slit performance hit. ![something](https://user-images.githubusercontent.com/13573860/35080588-8f56f3a0-fbdc-11e7-9ccf-7e5b095e3762.png)

Reply:
The "ModVersion" one is slightly faster. Is it the one with the suggested change?

Reply:
Sorry that's a little confusing. Here's a clearer image. 
![perf](https://user-images.githubusercontent.com/13573860/35103592-5141a8b4-fc34-11e7-9341-ed110bea6238.png)

Reply:
I found the following to be the fastest of the 4 ways that I tested:
```C++
int last_digit = y.to_string().back() - '0';
if (last_digit & 1)
    ...
```

Reply:
That should work, the operation that was causing the most performance loss was the modulo so we should be able to use this code. I'm about to test it now.

Reply:
It's still slightly slower on my machine.
![perf](https://user-images.githubusercontent.com/13573860/35107819-e7572120-fc3f-11e7-9895-7c1cdfeeb363.png)

Reply:
Dropping the `- '0'` will also have the same result but will be slightly faster.

Reply:
The results have been very strange. Sometimes it runs faster, sometimes it's the same speed and sometimes it's slower. I'm going to use a profiler to measure the performance because I don't trust the time command to be completely accurate.